Enforce CUDA >= 12 and fix its CMake search procedure

sogartar opened the pull request:
The IREE cuda2 driver depends on CUDA version >= 12. It uses the define CU_DEVICE_ATTRIBUTE_MEM_SYNC_DOMAIN_COUNT that appears in CUDA 12.

Also address a bug in CMake's procedure to find CUDA when the wrong system version is available.

ScottTodd commented on CMakeLists.txt:
```diff
@@ -195,7 +195,9 @@ option(IREE_BUILD_EXPERIMENTAL_WEB_SAMPLES "Builds experimental web samples." OF
 
 set(IREE_CUDA_AVAILABLE OFF)
-find_package(CUDAToolkit)
+# The IREE cuda2 driver requires CUDA >= 12.
```
Looks like this may fix https://github.com/openxla/iree/issues/14420 . I've context switched out some of the details there. Can you give that a read and see if anything is missing?

ScottTodd replied:
I don't think we need the "install our own version if the user provided a version that is too old" from that... that was just an idea to make things a bit easier for some users.

sogartar replied:
It is relevant. The solution in this PR is to
ignore the detected CUDA Toolkit and download our own version if the detected version is < 12.
This is one of the approaches you proposed there to handle an old CUDA version.

ScottTodd replied:
Ah, I see. Thanks!
(plus sorry for the delay, I'm working through a backlog of emails etc.)

Commit:
Enforce CUDA >= 12 and fix its CMake search procedure

The IREE cuda2 driver depends on CUDA version >= 12.
It uses the define CU_DEVICE_ATTRIBUTE_MEM_SYNC_DOMAIN_COUNT
that appears in CUDA 12.

Also address a bug in CMake's procedure to find CUDA when
the wrong system version is available.
